File: src/apps/shell.c
//
// Created by anerruption on 06/03/23.
//

#include "../../include/apps/shell.h"
#include "../../include/drivers/vga.h"
#include "../../include/drivers/ps2.h"
#include "../../include/drivers/acpi.h"
#include "../../include/utils/scan_map.h"
#include "../../include/memory/pmm.h"

size_t shell_read_line(char* buffer)
{
    size_t index = 0;

    for (;;)
    {
        asm("hlt");

        uint8_t scan_code = ps2_get_scan_code();

        if (scan_code == 0)
        {
            continue;
        }

        key keyboard_char = get_keyboard_key(scan_code);

        if (keyboard_char.type == UNKNOWN)
        {
            continue;
        }

        if (keyboard_char.type == BACKSPACE)
        {
            if (index > 0)
            {
                buffer[--index] = ' ';
                term_backspace();
            }
            continue;
        }

        if (keyboard_char.type == ENTER)
        {
            term_write_char('\n');
            return index;
        }

        buffer[index++] = keyboard_char.value;
        term_write_char(keyboard_char.value);
    }
}

void shell_exec()
{
    char* buffer = (char*)memory_alloc(1024);
    char* int_str = (char*)memory_alloc(16);

    size_t len;

    for (;;)
    {
        term_write_string("> ");

        int size = shell_read_line(buffer);

        if (is_equal(buffer, "help", size))
        {
            term_write_string("help - Shows all commands.\nclear - Clears the screen.\nusedram - Shows the amount of used memory, in KiB.\ntotalram - Shows the total amount of usable memory, in MiB.\nshutdown - Shuts down the PC via ACPI.\nreset - Resets the PC via ACPI.\n");
        }
        else if (is_equal(buffer, "clear", size))
        {
            term_clear();
        }
        else if (is_equal(buffer, "usedram", size))
        {
            term_write_string("RAM in use: ");
            len = itoa(get_pages_in_use() * get_page_size() / 1024, int_str, 10);
            term_write(int_str, len);
            term_write_string("K\n");
        }
        else if (is_equal(buffer, "totalram", size))
        {
            term_write_string("Total usable memory: ");
            len = itoa(get_total_usable_memory() / 1024 / 1024, int_str, 10);
            term_write(int_str, len);
            term_write_string("M\n");
        }
        else if (is_equal(buffer, "shutdown", size))
        {
            term_write_string("Shutting down...\n");
            acpi_shutdown();
        }
        else if (is_equal(buffer, "reset", size))
        {
            term_write_string("Resetting...\n");
            acpi_reset();
        }
        else
        {
            term_write_string("Unknown command: ");
            for (size_t i = 0; i < size; i++)
            {
                term_write_char(buffer[i]);
            }
            term_write_char('\n');
        }
    }
}

File: src/apps/shell.zig
const std = @import("std");
const vga = @import("../vga.zig");
const ps2 = @import("../ps2.zig");
const acpi = @import("../drivers/acpi.zig");
const pmm = @import("../memory/pmm.zig");
const scanmap = @import("../utils/scanmap.zig");

const BUFFER_SIZE = 4096;

var buffer: [BUFFER_SIZE]u8 = undefined;

fn read_line() usize {
    var index: usize = 0;

    while (true) {
        var scan_code = ps2.getScanCode();

        if (scan_code == 0) {
            continue;
        }

        var key = scanmap.getKey(scan_code);

        if (key.type == .unknown) {
            continue;
        }

        if (key.type == .backspace) {
            if (index > 0) {
                index -= 1;
                buffer[index] = ' ';
                vga.backspace();
            }

            continue;
        }

        if (key.type == .enter) {
            vga.newLine();
            return index;
        }

        buffer[index] = key.value;
        vga.putChar(key.value);

        index += 1;
    }
}

pub fn exec() void {
    const format_buffer_size = 1024;

    var format_buffer: [format_buffer_size]u8 = undefined;

    while (true) {
        vga.write("> ");

        var size = read_line();
        var command = buffer[0..size];

        if (std.mem.eql(u8, command, "help")) {
            vga.writeLine(
                \\help - Shows all commands.
                \\usedram - Shows the amount of used RAM, in KiB.
                \\totalram - Shows the total amount of usable RAM, in MiB.
                \\shutdown - Shuts down the computer via ACPI.
                \\reset - Resets the computer via ACPI.
            );
        } else if (std.mem.eql(u8, command, "clear")) {
            vga.clear();
        } else if (std.mem.eql(u8, command, "usedram")) {
            var format = std.fmt.bufPrint(&format_buffer, "RAM in use: {d} kiB", .{pmm.pages_in_use * pmm.PAGE_SIZE / 1024}) catch unreachable;
            vga.writeLine(format);
        } else if (std.mem.eql(u8, command, "totalram")) {
            var format = std.fmt.bufPrint(&format_buffer, "Total usable RAM: {d} MiB", .{pmm.total_size / 1024 / 1024}) catch unreachable;
            vga.writeLine(format);
        } else if (std.mem.eql(u8, command, "shutdown")) {
            vga.writeLine("Shutting down...");
            acpi.shutdown();
        } else if (std.mem.eql(u8, command, "reset")) {
            vga.writeLine("Resetting...");
            acpi.reset();
        } else {
            vga.write("Unknown command: ");
            vga.writeLine(command);
        }
    }
}
